src/charm.py

```diff
@@ -465,6 +469,25 @@ def _configure(self):  # noqa: C901
         new_layer = self._build_pebble_layer
         restart = current_layer.services != new_layer.services
 
+        # If v12_migration_date isn't set (due to missing or failed retrieval),
+        # we determine the migration date for v12 schema. This occurs once
+        # during initial setup, as subsequent hooks will get the value from the persisted backup config.
+
+        # If it's a fresh Loki installation, it's safe to set the v12 schema date to today.
+        # This ensures that logs are correctly managed in v12 from the beginning of Loki's operation.
+        # If it's an upgrade scenario, we set the date to tomorrow to accommodate today's logs
+        # that might have been written in the previous v11 schema format.
+
+        # By default, we assume it's a fresh installation unless state explicitly set to True
+        # by the upgrade hook, indicating an upgrade
+
+        if not (v12_migration_date := self.get_v12_migration_date_from_backup()):
+            today = datetime.date.today()
+            tomorrow = today + datetime.timedelta(days=1)
+            v12_migration_date = (tomorrow if self._stored.is_upgrade else today).strftime(
```

Comment:
Just though of a scenario we probably don't want to address, but could document in a comment?

If the pod is rescheduled during the startup sequence (e.g. due to manual kubectl delete or resource limit patch) - would we get the `upgrade-charm` event too early?